File: iosApp/iosApp/Presentation/Main/CafeManagement/MenuGoods/MenuGoodsUiState.swift
//
//  MenuGoodsUiState.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/11.
//

import Foundation
import Shared

struct MenuGoodsUiState {
    var cafeName: String = ""
    var isLoading: Bool = true
    var isSearchVisible: Bool = false
    var searchQuery: String = ""
    var selectedCollection: CollectionTab = .menu
    var menuCategories: [CategoryChip] = []
    var goodsCategories: [CategoryChip] = []
    var selectedMenuCategoryId: String? = nil
    var selectedGoodsCategoryId: String? = nil
    var menuItems: [CafeMenu] = []
    var goodsItems: [Goods] = []
    var menuAvailabilityOverrides: [String: Bool] = [:]
    var goodsAvailabilityOverrides: [String: Bool] = [:]
    var infoMessage: String? = nil
    var pendingDeleteItemId: String? = nil

    enum CollectionTab {
        case menu
        case goods
    }

    struct CategoryChip: Hashable {
        let id: String?
        let label: String
        let iconKey: String
    }

    var visibleCategories: [CategoryChip] {
        switch selectedCollection {
        case .menu:
            return menuCategories
        case .goods:
            return goodsCategories
        }
    }

    var selectedCategoryId: String? {
        switch selectedCollection {
        case .menu:
            return selectedMenuCategoryId
        case .goods:
            return selectedGoodsCategoryId
        }
    }

    func isMenuAvailable(_ menu: CafeMenu) -> Bool {
        menuAvailabilityOverrides[menu.id] ?? menu.isAvailable
    }

    func isGoodsAvailable(_ goods: Goods) -> Bool {
        goodsAvailabilityOverrides[goods.id] ?? (goods.stock > 0)
    }

    func menuCategoryId(_ menu: CafeMenu) -> String {
        menu.category.lowercased()
    }

    func menuCategoryLabel(_ menu: CafeMenu) -> String {
        switch menuCategoryId(menu) {
        case "food":
            return "Food"
        case "drink":
            return "Drinks"
        case "dessert":
            return "Dessert"
        default:
            return menu.category.capitalized
        }
    }

    func goodsCategoryId(_ goods: Goods) -> String {
        if goods.name.localizedCaseInsensitiveContains("포토") {
            return "collectible"
        } else if goods.name.localizedCaseInsensitiveContains("의상") {
            return "apparel"
        }
        return "goods"
    }

    func goodsCategoryLabel(_ goods: Goods) -> String {
        switch goodsCategoryId(goods) {
        case "collectible":
            return "Collectible"
        case "apparel":
            return "Apparel"
        default:
            return "Goods"
        }
    }

    var filteredMenuItems: [CafeMenu] {
        let normalizedQuery = searchQuery.trimmingCharacters(in: .whitespacesAndNewlines)
        return menuItems.filter { item in
            let matchesCategory = selectedMenuCategoryId == nil || menuCategoryId(item) == selectedMenuCategoryId
            let matchesQuery =
                normalizedQuery.isEmpty ||
                item.name.localizedCaseInsensitiveContains(normalizedQuery) ||
                item.desc.localizedCaseInsensitiveContains(normalizedQuery) ||
                menuCategoryLabel(item).localizedCaseInsensitiveContains(normalizedQuery)
            return matchesCategory && matchesQuery
        }
    }

    var filteredGoodsItems: [Goods] {
        let normalizedQuery = searchQuery.trimmingCharacters(in: .whitespacesAndNewlines)
        return goodsItems.filter { item in
            let matchesCategory = selectedGoodsCategoryId == nil || goodsCategoryId(item) == selectedGoodsCategoryId
            let matchesQuery =
                normalizedQuery.isEmpty ||
                item.name.localizedCaseInsensitiveContains(normalizedQuery) ||
                "카페 굿즈 판매 항목".localizedCaseInsensitiveContains(normalizedQuery) ||
                goodsCategoryLabel(item).localizedCaseInsensitiveContains(normalizedQuery)
            return matchesCategory && matchesQuery
        }
    }
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/cafemanagement/menugoods/MenuGoodsUiState.kt
package com.hhp227.concafe.presentation.main.cafemanagement.menugoods

import com.hhp227.concafe.domain.model.CafeMenu
import com.hhp227.concafe.domain.model.Goods

data class MenuGoodsUiState(
    val cafeName: String = "",
    val isLoading: Boolean = true,
    val isSearchVisible: Boolean = false,
    val searchQuery: String = "",
    val selectedCollection: CollectionTab = CollectionTab.MENU,
    val menuCategories: List<CategoryChip> = emptyList(),
    val goodsCategories: List<CategoryChip> = emptyList(),
    val selectedMenuCategoryId: String? = null,
    val selectedGoodsCategoryId: String? = null,
    val menuItems: List<CafeMenu> = emptyList(),
    val goodsItems: List<Goods> = emptyList(),
    val menuAvailabilityOverrides: Map<String, Boolean> = emptyMap(),
    val goodsAvailabilityOverrides: Map<String, Boolean> = emptyMap(),
    val infoMessage: String? = null,
    val pendingDeleteItemId: String? = null
) {
    enum class CollectionTab {
        MENU,
        GOODS
    }

    data class CategoryChip(
        val id: String?,
        val label: String,
        val iconKey: String
    )

    val visibleCategories: List<CategoryChip>
        get() = when (selectedCollection) {
            CollectionTab.MENU -> menuCategories
            CollectionTab.GOODS -> goodsCategories
        }

    val selectedCategoryId: String?
        get() = when (selectedCollection) {
            CollectionTab.MENU -> selectedMenuCategoryId
            CollectionTab.GOODS -> selectedGoodsCategoryId
        }

    fun isMenuAvailable(menu: CafeMenu): Boolean {
        return menuAvailabilityOverrides[menu.id] ?: menu.isAvailable
    }

    fun isGoodsAvailable(goods: Goods): Boolean {
        return goodsAvailabilityOverrides[goods.id] ?: (goods.stock > 0)
    }

    fun menuCategoryId(menu: CafeMenu): String = menu.category.lowercase()

    fun menuCategoryLabel(menu: CafeMenu): String {
        return when (menuCategoryId(menu)) {
            "food" -> "Food"
            "drink" -> "Drinks"
            "dessert" -> "Dessert"
            else -> menu.category.replaceFirstChar { it.uppercase() }
        }
    }

    fun goodsCategoryId(goods: Goods): String {
        return when {
            goods.name.contains("포토", ignoreCase = true) -> "collectible"
            goods.name.contains("의상", ignoreCase = true) -> "apparel"
            else -> "goods"
        }
    }

    fun goodsCategoryLabel(goods: Goods): String {
        return when (goodsCategoryId(goods)) {
            "collectible" -> "Collectible"
            "apparel" -> "Apparel"
            else -> "Goods"
        }
    }

    val filteredMenuItems: List<CafeMenu>
        get() {
            val normalizedQuery = searchQuery.trim()
            return menuItems.filter { item ->
                val matchesCategory = selectedMenuCategoryId == null || menuCategoryId(item) == selectedMenuCategoryId
                val matchesQuery = normalizedQuery.isBlank() ||
                    item.name.contains(normalizedQuery, ignoreCase = true) ||
                    item.desc.contains(normalizedQuery, ignoreCase = true) ||
                    menuCategoryLabel(item).contains(normalizedQuery, ignoreCase = true)
                matchesCategory && matchesQuery
            }
        }

    val filteredGoodsItems: List<Goods>
        get() {
            val normalizedQuery = searchQuery.trim()
            return goodsItems.filter { item ->
                val matchesCategory = selectedGoodsCategoryId == null || goodsCategoryId(item) == selectedGoodsCategoryId
                val matchesQuery = normalizedQuery.isBlank() ||
                    item.name.contains(normalizedQuery, ignoreCase = true) ||
                    "카페 굿즈 판매 항목".contains(normalizedQuery, ignoreCase = true) ||
                    goodsCategoryLabel(item).contains(normalizedQuery, ignoreCase = true)
                matchesCategory && matchesQuery
            }
        }
}
